rfd/xxxx-headless-authentication.md

```diff
@@ -0,0 +1,453 @@
+    webauthn.CredentialAssertionResponse Webauthn = 3;
+  }
+}
+
+
+// UpdateHeadlessLoginRequestStateResponse is a response to update a headless
+// login request's state. Denial requests should return an empty response, since
+// MFA challenge/response flow is not needed.
+message UpdateHeadlessLoginRequestStateResponse {
+  // MFAChallenge is an MFA challenge for a user's registered MFA devices
+  MFAAuthenticateChallenge MFAChallenge = 1;
+}
+
+// MFAAuthenticateChallenge is a challenge for all MFA devices registered for a user.
+message MFAAuthenticateChallenge {
+  reserved 1; // repeated U2FChallenge U2F
+  // TOTP is a challenge for all TOTP devices registered for a user. When
+  // this field is set, any TOTP device a user has registered can be used to
+  // respond.
+  TOTPChallenge TOTP = 2;
+  // WebauthnChallenge contains a Webauthn credential assertion used for
+  // login/authentication ceremonies.
+  // Credential assertions hold, among other information, a list of allowed
+  // credentials for the ceremony (one for each U2F or Webauthn device
+  // registered by the user).
+  webauthn.CredentialAssertion WebauthnChallenge = 3;
+}
+```
+
+### UX
+
+#### `tsh --headless`
+
+We will add a new `--headless` flag to `tsh` which can be used to authenticate for a single `tsh` request. When this flag is provided, `tsh` will prompt the user to complete headless authentication on their local machine with the command `tsh headless approve <request_id>` or the URL `https://proxy.example.com/headless/<request_id>/approve`. Once the user completes local authentication, `tsh` will receive credentials to complete the request.
+
+```console
+$ tsh --headless --proxy=proxy --user=user ssh user@node01
+Complete headless authentication on your local device or web browser:
+ - tsh --proxy=proxy.example.com --user=dev headless approve <request_id>
+ - https://proxy.example.com/headless/<request_id>/approve
+// Wait for user to complete local authentication with MFA
+<user@node01> $
+```
+
+#### Environment variables
+
+In the `tsh --headless` flow, users never run `tsh login` on their remote machine. Instead, we expect the `--proxy`, `--user`, and `--headless` flags to be supplied to each command. To reduce UX friction, users can set the environment variables `TELEPORT_PROXY=<proxy_addr>`, `TELEPORT_USER=<user>`, and `TELEPORT_HEADLESS=true` instead.
+
+We prefer setting environment variables rather than saving config to disk (`~/.tsh/proxy.example.com.yaml` and `~/.tsh/current-profile`) so that the headless flow remains stateless, preventing conflicts on shared machines.
+
+#### `tsh headless`
+
+When the user enters the headless authentication command or URL, the user will be prompted to login with MFA, if they are not logged in already. The user will then be notified of additional request details and asked to acknowledge that they made this request (`y/N` prompt). Finally, the user is asked to verify with MFA to approve the request.
```

Comment:
How is RBAC to these requests controlled? I don't think I've seen it addressed above. For example, who can approve/deny these requests? Can users only approve their own requests or can admins see/approve others' requests too via some resource rules (similar to access requests' RBAC)?

Reply:
`UpdateHeadlessAuthRequestState` will only be authorized for the requesting Teleport user. The latter will also require the user's MFA verification.